// @tonaljs/tonal/dist/index.es.js
export { default as AbcNotation } from '@tonaljs/abc-notation';
import * as array from '@tonaljs/array';
export { array as Array };
export { default as Chord } from '@tonaljs/chord';
import ChordType from '@tonaljs/chord-type';
export { default as ChordType } from '@tonaljs/chord-type';
export { default as Collection } from '@tonaljs/collection';
import * as Core from '@tonaljs/core';
export * from '@tonaljs/core';
export { Core };
export { default as DurationValue } from '@tonaljs/duration-value';
export { default as Interval } from '@tonaljs/interval';
export { default as Key } from '@tonaljs/key';
export { default as Midi } from '@tonaljs/midi';
export { default as Mode } from '@tonaljs/mode';
export { default as Note } from '@tonaljs/note';
import Pcset from '@tonaljs/pcset';
export { default as Pcset } from '@tonaljs/pcset';
export { default as Progression } from '@tonaljs/progression';
export { default as Range } from '@tonaljs/range';
export { default as RomanNumeral } from '@tonaljs/roman-numeral';
export { default as Scale } from '@tonaljs/scale';
import ScaleType from '@tonaljs/scale-type';
export { default as ScaleType } from '@tonaljs/scale-type';
export { default as TimeSignature } from '@tonaljs/time-signature';

// deprecated (backwards compatibility)
const Tonal = Core;
const PcSet = Pcset;
const ChordDictionary = ChordType;
const ScaleDictionary = ScaleType;

export { ChordDictionary, PcSet, ScaleDictionary, Tonal };
